### PR TITLE
network: make IgnoreCarrierLoss= also take a timespan

### LINKED ISSUE #18738: systemd-networkd loops in gaining/losing carrier with dhcp-provided mtu for some nics
**systemd version the issue has been seen with**
 > latest git

**Used distribution**
 > Ubuntu

**Expected behaviour you didn't see**
 > interface configured for DHCPv4 with server-provided MTU comes up successfully

**Unexpected behaviour you saw**
 > interface remains in loop toggling between carrier up/down, constantly setting the mtu to dhcp-provided mtu and original mtu

**Steps to reproduce the problem**
 > This requires a nic that drops carrier when setting its mtu; the e1000e driver/hw is one such nic.
> 1) create a VM and configure to use e1000e emulation for its interface
> 2) connect that VM to a network that provides DHCPv4
> 3) configure the DHCPv4 server to provide a non-standard (i.e. not 1500) mtu
> 4) start up the test VM and check systemd-networkd logs
> 5) the logs will show systemd-networkd cycling endlessly between carrier up and down
> As this is a timing issue, it's possible the interface will eventually successfully come up.

When the problem is reproduced, systemd-networkd will show logs similar to:

> Feb 22 15:18:58 lp1881207-upstream systemd-networkd[992]: ens8: Gained carrier
> Feb 22 15:18:58 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv4 address 1.2.3.133/24 via 1.2.3.1
> Feb 22 15:18:59 lp1881207-upstream systemd-networkd[992]: ens8: Lost carrier
> Feb 22 15:18:59 lp1881207-upstream systemd-networkd[992]: ens8: DHCP lease lost
> Feb 22 15:18:59 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv6 lease lost
> Feb 22 15:19:01 lp1881207-upstream systemd-networkd[992]: ens8: Gained carrier
> Feb 22 15:19:01 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv4 address 1.2.3.133/24 via 1.2.3.1
> Feb 22 15:19:02 lp1881207-upstream systemd-networkd[992]: ens8: Lost carrier
> Feb 22 15:19:02 lp1881207-upstream systemd-networkd[992]: ens8: DHCP lease lost
> Feb 22 15:19:02 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv6 lease lost
> Feb 22 15:19:04 lp1881207-upstream systemd-networkd[992]: ens8: Gained carrier
> Feb 22 15:19:04 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv4 address 1.2.3.133/24 via 1.2.3.1
> Feb 22 15:19:05 lp1881207-upstream systemd-networkd[992]: ens8: Lost carrier
> Feb 22 15:19:05 lp1881207-upstream systemd-networkd[992]: ens8: DHCP lease lost
> Feb 22 15:19:05 lp1881207-upstream systemd-networkd[992]: ens8: DHCPv6 lease lost
> 

This is not a new problem, and the code does currently work around this by setting link->setting_mtu flag while the mtu netlink request is pending; however as soon as the kernel sends the netlink response, systemd-networkd clears the link->setting_mtu flag.

When using ip monitor to see the timing, it looks like the kernel first reports the MTU change, and then less than 1ms later the kernel reports the carrier loss. I'm not sure if this ordering (mtu change first, then carrier loss) is new behavior from the kernel, or if it's always been this way.

> root@lp1881207-upstream:~# ip -4 l show dev ens8 ; ip -4 -t monitor l dev ens8
> 3: ens8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
>     link/ether 52:54:00:15:2c:8a brd ff:ff:ff:ff:ff:ff
> Timestamp: Mon Feb 22 19:57:54 2021 1749 usec
> [LINK]3: ens8: <BROADCAST,MULTICAST,UP> mtu 9000 qdisc fq_codel state UP group default 
>     link/ether 52:54:00:15:2c:8a brd ff:ff:ff:ff:ff:ff
> Timestamp: Mon Feb 22 19:57:54 2021 3852 usec
> [LINK]3: ens8: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 9000 qdisc fq_codel state DOWN group default 
>     link/ether 52:54:00:15:2c:8a brd ff:ff:ff:ff:ff:ff
> Timestamp: Mon Feb 22 19:57:56 2021 226630 usec
> [LINK]3: ens8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc fq_codel state UP group default 
>     link/ether 52:54:00:15:2c:8a brd ff:ff:ff:ff:ff:ff
> 

After that monitor was started, I changed the mtu to 9000, and the ordering of the netlink events shows the mtu is set first, then the carrier is dropped.

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #18738 and #20887.

Replaces #18746.

### CODE REVIEW

**Comment by yuwata** [comment]
@ddstreet and @zeldovich Could you test this PR?

For issue #20887, I tested `IgnoreCarrierLoss=3sec` with restarting wpa_supplicant.service, and it works fine!

For issue #18738, unfortunately I do not have such hardware, so I cannot test...

**Comment by zeldovich** [comment]
This works well for addressing the wifi roaming use case (issue #20887); thank you!

**Comment by yuwata** [comment]
@zeldovich founds an issue on this PR. See https://github.com/systemd/systemd/issues/20887#issuecomment-968052010. Setting the 'dont-merge' label.

**Comment by yuwata** [comment]
@zeldovich The reported issue should be fixed now. Could you test one more time? Thank you!

**Comment by zeldovich** [comment]
Looks good to me.  The latest version of this PR behaves correctly with respect to wireless roaming both between APs in the same SSID and across SSIDs.  Thank you!

File: man/systemd.network.xml [hunk 1]
```diff
@@ -952,18 +952,21 @@ Table=1234</programlisting></para>
...
       <varlistentry>
         <term><varname>IgnoreCarrierLoss=</varname></term>
         <listitem>
-          <para>Takes a boolean. Allows networkd to retain both the static and dynamic configuration
-          of the interface even if its carrier is lost. When unset, the value specified with
-          <option>ConfigureWithoutCarrier=</option> is used.
-          </para>
+          <para>Takes a boolean or a timespan. When true, networkd retains both the static and dynamic
+          configuration of the interface even if its carrier is lost. When a timespan is specified,
+          networkd wait for the specified timespan, and ignore the carrier loss if the link regain its
```
Comment (bluca) [inline man/systemd.network.xml]: "waits" and "ignores"

File: man/systemd.network.xml [hunk 2]
```diff
@@ -952,18 +952,21 @@ Table=1234</programlisting></para>
         <term><varname>ConfigureWithoutCarrier=</varname></term>
         <listitem>
           <para>Takes a boolean. Allows networkd to configure a specific link even if it has no carrier.
-          Defaults to false. If <option>IgnoreCarrierLoss=</option> is not explicitly set, it will
-          default to this value.
+          Defaults to false. If enabled, and the <varname>IgnoreCarrierLoss=</varname> setting is not
+          explicitly set, then it is enabled.
```
Comment (bluca) [inline man/systemd.network.xml]: maybe "then it is enabled as well" to indicate that it's talking about the second setting

File: man/systemd.network.xml [hunk 3]
```diff
@@ -1805,6 +1808,9 @@ Table=1234</programlisting></para>
             <para>When true, the interface maximum transmission unit from the DHCP server will be used on the
             current link. If <varname>MTUBytes=</varname> is set, then this setting is ignored. Defaults to
             false.</para>
+            <para>Note, some drivers reset interfaces on changing MTU. For such interfaces, please try
```
Comment (bluca) [inline man/systemd.network.xml]: "some drivers will reset the interfaces if the MTU is changed"

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -1580,16 +1584,30 @@ int manager_udev_process_link(sd_device_monitor *monitor, sd_device *device, voi
...
+        bool force_reconfigure;
         int r;
 
         assert(link);
 
+        r = event_source_disable(link->carrier_lost_timer);
+        if (r < 0)
+                log_link_warning_errno(link, r, "Failed to disable carrier lost timer, ignoring: %m");
+
+        /* If the SSID is changed, then the connected wireless network should be changed. So, always
```
Comment (bluca) [inline src/network/networkd-link.c]: Should be changed? Did you mean "could be changed" here?

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -1615,6 +1633,45 @@ static int link_carrier_gained(Link *link) {
...
+        return ret;
+}
+
+static int link_carrier_lost_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        Link *link = userdata;
+        int r;
+
+        assert(link);
+
+        r = link_carrier_lost_impl(userdata);
```
Comment (bluca) [inline src/network/networkd-link.c]: doesn't change anything, but maybe pass `link` here given you have it, as it's a bit clearer

File: src/network/networkd-link.c [hunk 6]
```diff
@@ -1580,16 +1584,30 @@ int manager_udev_process_link(sd_device_monitor *monitor, sd_device *device, voi
...
         int r;
 
         assert(link);
 
+        r = event_source_disable(link->carrier_lost_timer);
+        if (r < 0)
+                log_link_warning_errno(link, r, "Failed to disable carrier lost timer, ignoring: %m");
+
+        /* If the SSID is changed, then the connected wireless network should be changed. So, always
+         * reconfigure the link. That makes e.g. DHCP client being restarted, and the correct network
```
Comment (bluca) [inline src/network/networkd-link.c]: maybe: "Which means the DHCP client will be restarted, ..."

**Review by bluca — APPROVED** [review]
LGTM, just some minor nits

**Comment by yuwata** [comment]
@bluca Thank you for always correcting my bad English. All above points are addressed. Upgrading the green label.

**Pull request merged**